python/ml4ir/base/model/architectures/auto_dag_network.py

```diff
@@ -0,0 +1,362 @@
+
+        self.file_io: FileIO = file_io
+        self.model_config = model_config
+        self.feature_config = feature_config
+
+        # Get all available layers (including keras layers)
+        self.available_layers = get_layer_subclasses()
+        model_graph = self.define_architecture(model_config)
+        self.execution_order: List[LayerNode] = model_graph.topological_sort()
+        # The line below is important for tensorflow to register the available params for the model
+        # An alternative is to do this in build()
+        # If removed, no layers will be present in the AutoDagNetwork (in the model summary)
+        self.register_layers: List[layers.Layer] = [layer_node.layer for layer_node in self.execution_order if
+                                                    not layer_node.is_input_node]
+        self.file_io.logger.info("Execution order: %s", self.execution_order)
+        self.output_node = model_graph.output_node
+
+    def instantiate_op(self, current_layer_type: str, current_layer_args: Dict) -> layers.Layer:
+        """
+        Create and return an instance of `current_layer_type` with `current_layer_args` params
+        If `current_layer_type` is not found in the set of subclasses of keras.layers.Layer, throws KeyError
+
+        Parameters
+        ----------
+        current_layer_type: str
+            Refers to class name inheriting directly or indirectly from keras.layers.Layer to be instantiated
+        current_layer_args: dict
+            All the arguments from model config which are used as kwargs to current_layer_type instance
+
+        Returns
+        -------
+        keras.layers.Layer
+            Instance of current_layer_type
+        """
+        try:
+            return self.available_layers[current_layer_type](**current_layer_args)
+        except KeyError:
+            self.file_io.logger.error("Layer type: '%s' is not supported or not found in %s",
+                                      current_layer_type, str(self.available_layers))
+            raise KeyError(f"Layer type: '{current_layer_type}' "
+                           f"is not supported or not found in subclasses of keras.layers.Layer")
+
+    def get_layer_op(self, layer_args: Dict) -> Dict[str, Union[str, bool, layers.Layer]]:
+        """
+        Define the layer operation for a layer in the model config
+
+        Parameters
+        ----------
+        layer_args: dict
+            All the arguments from model config which are needed to define a layer op
+
+        Returns
+        -------
+        dict
+            Dictionary with the layer instance, inputs required for the layer and the format of inputs to the layer
+        """
+        return {
+            self.INPUTS: layer_args[self.INPUTS],
+            self.OP_IDENTIFIER: self.instantiate_op(layer_args[self.LAYER_TYPE], layer_args.get(self.LAYER_KWARGS, {})),
+            self.INPUTS_AS_LIST: layer_args.get(self.INPUTS_AS_LIST, False)
```

Comment:
@arvindsrikantan I'm still not very clear on why we need `as_list`.

Reply:
Layers can take 3 forms of inputs -> tensor, dict, list. If there is only 1 input to a node, we automatically choose to only send the tensor. If the layer accepts multiple inputs, this flag controls what the format of the input should be (dict being the default). Let me know if I'm missing something